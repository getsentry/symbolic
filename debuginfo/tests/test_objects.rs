use std::fmt;

use failure::Error;
use insta;

use symbolic_common::ByteView;
use symbolic_debuginfo::{DebugSession, Function, Object, SymbolMap};

/// Helper to create neat snapshots for symbol tables.
struct SymbolsDebug<'a>(&'a SymbolMap<'a>);

impl fmt::Debug for SymbolsDebug<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for symbol in self.0.iter() {
            writeln!(
                f,
                "{:>16x} {}",
                &symbol.address,
                &symbol.name().unwrap_or("<unknown>")
            )?;
        }

        Ok(())
    }
}

/// Helper to create neat snapshots for function trees.
struct FunctionsDebug<'a>(&'a [Function<'a>], usize);

impl fmt::Debug for FunctionsDebug<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for function in self.0 {
            writeln!(
                f,
                "\n{:indent$}> {:#x}: {} ({:#x})",
                "",
                function.address,
                function.name,
                function.size,
                indent = self.1 * 2
            )?;

            for line in &function.lines {
                writeln!(
                    f,
                    "{:indent$}  {:#x}: {}:{} ({})",
                    "",
                    line.address,
                    line.file.name_str(),
                    line.line,
                    line.file.dir_str(),
                    indent = self.1 * 2
                )?;
            }

            write!(f, "{:?}", FunctionsDebug(&function.inlinees, self.1 + 1))?;
        }

        Ok(())
    }
}

#[test]
fn test_breakpad() -> Result<(), Error> {
    // Using the windows version here since it contains all record kinds
    let view = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Breakpad(
    BreakpadObject {
        code_id: Some(
            CodeId(5ab380779000)
        ),
        debug_id: DebugId {
            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
            appendix: 1
        },
        arch: X86,
        name: "crash.pdb",
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_breakpad_symbols() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("breakpad_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_breakpad_functions() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!(
        "breakpad_functions",
        FunctionsDebug(&functions[..10], 0)
    );

    Ok(())
}

#[test]
fn test_elf_executable() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/linux/crash")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Elf(
    ElfObject {
        code_id: Some(
            CodeId(f1c3bcc0279865fe3058404b2831d9e64135386c)
        ),
        debug_id: DebugId {
            uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
            appendix: 0
        },
        arch: Amd64,
        kind: Executable,
        load_address: 0x400000,
        has_symbols: true,
        has_debug_info: false,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_elf_debug() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/linux/crash.debug")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Elf(
    ElfObject {
        code_id: Some(
            CodeId(f1c3bcc0279865fe3058404b2831d9e64135386c)
        ),
        debug_id: DebugId {
            uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
            appendix: 0
        },
        arch: Amd64,
        kind: Debug,
        load_address: 0x400000,
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: false
    }
)"###);

    Ok(())
}

#[test]
fn test_elf_symbols() -> Result<(), Error> {
    // TODO(ja): Why does crash.debug not retain the symbol table but report has_symbols
    let view = ByteView::open("../testutils/fixtures/linux/crash")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("elf_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_elf_functions() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/linux/crash.debug")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!("elf_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}

#[test]
fn test_mach_executable() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/macos/crash")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"MachO(
    MachObject {
        code_id: Some(
            CodeId(67e9247c814e392ba027dbde6748fcbf)
        ),
        debug_id: DebugId {
            uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
            appendix: 0
        },
        arch: Amd64,
        kind: Executable,
        load_address: 0x100000000,
        has_symbols: true,
        has_debug_info: false,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_mach_dsym() -> Result<(), Error> {
    let view =
        ByteView::open("../testutils/fixtures/macos/crash.dSYM/Contents/Resources/DWARF/crash")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"MachO(
    MachObject {
        code_id: Some(
            CodeId(67e9247c814e392ba027dbde6748fcbf)
        ),
        debug_id: DebugId {
            uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
            appendix: 0
        },
        arch: Amd64,
        kind: Debug,
        load_address: 0x100000000,
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: false
    }
)"###);

    Ok(())
}

#[test]
fn test_mach_symbols() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/macos/crash")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("mach_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_mach_functions() -> Result<(), Error> {
    let view =
        ByteView::open("../testutils/fixtures/macos/crash.dSYM/Contents/Resources/DWARF/crash")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!("mach_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}

#[test]
fn test_pe_32() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.exe")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Pe(
    PeObject {
        code_id: Some(
            CodeId(5ab380779000)
        ),
        debug_id: DebugId {
            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
            appendix: 1
        },
        debug_file_name: Some(
            "C:\\projects\\breakpad-tools\\windows\\Release\\crash.pdb"
        ),
        arch: X86,
        kind: Executable,
        load_address: 0x400000,
        has_symbols: false,
        has_debug_info: false,
        has_unwind_info: false
    }
)"###);

    Ok(())
}

#[test]
fn test_pe_64() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/CrashWithException.exe")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Pe(
    PeObject {
        code_id: Some(
            CodeId(5c9e09599000)
        ),
        debug_id: DebugId {
            uuid: "f535c5fb-2ae8-4bb8-aa20-6c30be566c5a",
            appendix: 1
        },
        debug_file_name: Some(
            "C:\\Users\\sentry\\source\\repos\\CrashWithException\\x64\\Release\\CrashWithException.pdb"
        ),
        arch: Amd64,
        kind: Executable,
        load_address: 0x140000000,
        has_symbols: false,
        has_debug_info: false,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

// NB: No test for PE symbols because our executable does not export any symbols
// NB: No test for PE functions because we can only read debug info from PDBs

#[test]
fn test_pdb() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Pdb(
    PdbObject {
        debug_id: DebugId {
            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
            appendix: 1
        },
        arch: X86,
        load_address: 0x0,
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_pdb_symbols() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("pdb_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_pdb_functions() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!("pdb_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}
