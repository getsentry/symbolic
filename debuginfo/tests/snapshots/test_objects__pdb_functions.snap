---
created: "2019-04-10T12:21:36.942698Z"
creator: insta@0.7.4
source: debuginfo/tests/test_objects.rs
expression: "FunctionsDebug(&functions[..10], 0)"
---

> 0x1000: ?RequestDump@CrashGenerationClient@google_breakpad@@QAE_NPAU_EXCEPTION_POINTERS@@PAUMDRawAssertionInfo@@@Z (0x114)
  0x1000: crash_generation_client.cc:323 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\crash_generation)
  0x1011: crash_generation_client.cc:324 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\crash_generation)
  0x1017: crash_generation_client.cc:325 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\crash_generation)
  0x101a: crash_generation_client.cc:338 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\crash_generation)
  0x102b: crash_generation_client.cc:328 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\crash_generation)
  0x1033: crash_generation_client.cc:329 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\crash_generation)
  0x1039: crash_generation_client.cc:331 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\crash_generation)
  0x1047: crash_generation_client.cc:332 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\crash_generation)
  0x1052: crash_generation_client.cc:333 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\crash_generation)
  0x1054: crash_generation_client.cc:334 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\crash_generation)
  0x1064: crash_generation_client.cc:337 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\crash_generation)
  0x10fd: crash_generation_client.cc:338 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\crash_generation)
  0x1101: crash_generation_client.cc:337 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\crash_generation)
  0x1107: crash_generation_client.cc:338 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\crash_generation)

> 0x1120: ??1?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAE@XZ (0x54)
  0x1120: xstring:2424 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x1123: xstring:2425 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x116c: xstring:2426 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x116d: xstring:2425 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)

> 0x1180: ?assign@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAEAAV12@QB_WI@Z (0x150)
  0x1180: xstring:2612 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x1195: xstring:2614 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x11a0: xstring:2616 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x11a9: xstring:2618 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x11b0: xstring:2617 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x11b3: xstring:2618 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x11bf: xstring:2619 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x11c5: xstring:2623 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x11c8: xstring:2627 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x11d1: xstring:2623 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x12c1: xstring:2627 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x12ca: xstring:2623 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)

> 0x12d0: ?_Xlen@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@SAXXZ (0xc)
  0x12d0: xstring:3918 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x12d0: xstring:3919 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)

> 0x12e0: ??4?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAEAAV01@$$QAV01@@Z (0x7f)
  0x12e0: xstring:2267 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x12e8: xstring:2268 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x12ec: xstring:2270 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x1334: xstring:2274 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x1352: xstring:2277 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x1354: xstring:2278 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)
  0x1358: xstring:2270 (c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.13.26128\include)

> 0x1360: __local_stdio_printf_options (0x6)
  0x1360: corecrt_stdio_config.h:85 (c:\program files (x86)\windows kits\10\include\10.0.16299.0\ucrt)
  0x1360: corecrt_stdio_config.h:87 (c:\program files (x86)\windows kits\10\include\10.0.16299.0\ucrt)
  0x1365: corecrt_stdio_config.h:88 (c:\program files (x86)\windows kits\10\include\10.0.16299.0\ucrt)

> 0x1370: swprintf_s (0x35)
  0x1370: corecrt_wstdio.h:1538 (c:\program files (x86)\windows kits\10\include\10.0.16299.0\ucrt)
  0x1379: corecrt_wstdio.h:1542 (c:\program files (x86)\windows kits\10\include\10.0.16299.0\ucrt)
  0x13a1: corecrt_wstdio.h:1545 (c:\program files (x86)\windows kits\10\include\10.0.16299.0\ucrt)

> 0x13b0: _snwprintf_s (0x34)
  0x13b0: corecrt_wstdio.h:1723 (c:\program files (x86)\windows kits\10\include\10.0.16299.0\ucrt)
  0x13b4: corecrt_wstdio.h:1727 (c:\program files (x86)\windows kits\10\include\10.0.16299.0\ucrt)
  0x13e3: corecrt_wstdio.h:1730 (c:\program files (x86)\windows kits\10\include\10.0.16299.0\ucrt)

> 0x13f0: fprintf (0x24)
  0x13f0: stdio.h:833 (c:\program files (x86)\windows kits\10\include\10.0.16299.0\ucrt)
  0x13f4: stdio.h:837 (c:\program files (x86)\windows kits\10\include\10.0.16299.0\ucrt)
  0x1413: stdio.h:840 (c:\program files (x86)\windows kits\10\include\10.0.16299.0\ucrt)

> 0x1420: ?Initialize@ExceptionHandler@google_breakpad@@AAEXABV?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@P6A_NPAXPAU_EXCEPTION_POINTERS@@PAUMDRawAssertionInfo@@@ZP6A_NPB_W5123_N@Z1HW4_MINIDUMP_TYPE@@51PAVCrashGenerationClient@2@PBUCustomClientInfo@2@@Z (0x34a)
  0x1420: exception_handler.cc:148 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1450: exception_handler.cc:149 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x145e: exception_handler.cc:150 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1464: exception_handler.cc:151 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x146b: exception_handler.cc:152 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1472: exception_handler.cc:153 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1479: exception_handler.cc:154 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1480: exception_handler.cc:155 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1487: exception_handler.cc:156 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x148e: exception_handler.cc:157 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1495: exception_handler.cc:158 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x149c: exception_handler.cc:159 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x14a3: exception_handler.cc:160 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x14aa: exception_handler.cc:161 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x14b1: exception_handler.cc:162 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x14b8: exception_handler.cc:164 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x14c2: exception_handler.cc:166 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x14cc: exception_handler.cc:167 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x14d9: exception_handler.cc:168 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x14e0: exception_handler.cc:169 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x14ea: exception_handler.cc:170 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x14f4: exception_handler.cc:171 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x14fe: exception_handler.cc:172 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1508: exception_handler.cc:173 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1512: exception_handler.cc:174 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x151b: exception_handler.cc:176 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1522: exception_handler.cc:180 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1530: exception_handler.cc:199 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x153a: exception_handler.cc:209 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1547: exception_handler.cc:210 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1555: exception_handler.cc:211 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1579: exception_handler.cc:213 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x158d: exception_handler.cc:214 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x15a5: exception_handler.cc:217 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x15b7: exception_handler.cc:220 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x15d6: exception_handler.cc:226 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x15ee: exception_handler.cc:229 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x15f9: exception_handler.cc:230 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1606: exception_handler.cc:231 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1611: exception_handler.cc:238 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x161f: exception_handler.cc:239 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1623: exception_handler.cc:240 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x162e: exception_handler.cc:246 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x165c: exception_handler.cc:253 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1669: exception_handler.cc:252 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1675: exception_handler.cc:253 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x169f: exception_handler.cc:266 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x16a3: exception_handler.cc:253 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x16ae: exception_handler.cc:266 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x16b0: exception_handler.cc:267 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x16bb: exception_handler.cc:271 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x16c6: exception_handler.cc:275 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x16cf: exception_handler.cc:276 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1700: exception_handler.cc:278 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x170e: exception_handler.cc:281 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1719: exception_handler.cc:285 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1727: exception_handler.cc:289 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x1741: exception_handler.cc:291 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)
  0x174c: exception_handler.cc:293 (c:\projects\breakpad-tools\deps\breakpad\src\client\windows\handler)

